File: leetcode/2316.统计无向图中无法互相到达点对数.py
class Solution:
    def countPairs(self, n: int, edges: List[List[int]]) -> int:

        g = [[] * n for _ in  range(n)]
        for x, y in edges:
            g[x].append(y)
            g[y].append(x)
        # print(g)
        def dfs(x):
            res = 1 # 本身自己这个点
            vis[x] = True
            for y in g[x]:
                if vis[y]: continue
                res += dfs(y)
            return res
        vis = [False] * n
        ans = []
        for i in range(n):
            if not vis[i]:
                ans.append(dfs(i))
                # print(i, vis)
        # print(ans)
        res = 0
        s = 0
        for i, x in enumerate(ans):
            res += x * s
            s += x
        return res

        

File: leetcode/2316.统计无向图中无法互相到达点对数.go
func countPairs(n int, edges [][]int) int64 {
    g := make([][]int, n)
    vis := make([]bool, n)
    ans := []int{}
    for i := 0; i < n; i++ {
        g[i] = []int{}
    }
    for _, edge := range edges {
        x, y := edge[0], edge[1]
        g[x] = append(g[x], y)
        g[y] = append(g[y], x)
    }
    var dfs func(int) int
    dfs = func(x int) int {
        res := 1
        vis[x] = true
        for _, y := range g[x] {
            if vis[y]{
                continue
            }
            res += dfs(y)
        }
        return res
    }
    for i := 0; i < n; i++ {
        if !vis[i] {
            ans = append(ans, dfs(i))
        }
    }
    res := 0
    s := 0
    for _, x := range ans {
        res += x * s
        s += x
    }
    return int64(res)
}